File: services/source_service.py
import json
from pathlib import Path

from models.song import Song
from models.audio import Audio
from models.band import Band


class MissingAudioFilesError(FileNotFoundError):

    def __init__(self, missing_files):
        self.missing_files = missing_files
        message = "Audio files not found:\n{}".format(
            "\n".join(str(file) for file in missing_files)
        )
        super().__init__(message)


class SourceService:

    def __init__(self, config_file='source.json'):

        self._data = None
        self._config_file = Path(config_file)
        self._project_root = self._config_file.resolve().parent
        self._load()

    def bands(self):

        bands = []

        for band in self._data['bands']:
            bands.append(Band(band['id'], band['name'], band['logo']))

        return bands

    def songs(self, band: Band):

        songs = []

        path = self._resolve_path(self._data.get('path', 'bands'))

        for song in self._data[band.id]:

            audios = []

            for audio in song['audios']:

                file = self._path_audio_file(path, band, song['id'], audio['file'])

                audios.append(
                    Audio(audio['file'], audio['name'], str(file)))

            songs.append(Song(
                song['id'], song['name'], song['autoforward'], audios))
        
        
        return songs

    def validate_audio_files(self):
        missing_files = []

        for band in self.bands():
            for song in self.songs(band):
                for audio in song.audios:
                    audio_file = Path(audio.file)

                    if not audio_file.is_file():
                        missing_files.append(audio_file)

        if len(missing_files) > 0:
            raise MissingAudioFilesError(missing_files)


    def _load(self):
        with self._config_file.open('r', encoding='utf-8-sig') as json_file:
            self._data = json.load(json_file)

    def _resolve_path(self, path: str):
        source_path = Path(path)

        if source_path.is_absolute():
            return source_path

        return self._project_root / source_path

    def _path_audio_file(self, path: Path, band: Band, song_id: str, file: str):
        return path / band.id / song_id / file


Source_Service = SourceService

File: native/src/source_service.cpp
#include "source_service.hpp"

#include "json_file_reader.hpp"

namespace cherry {

namespace {

std::filesystem::path project_root_of(const std::filesystem::path& config_file) {
    return std::filesystem::absolute(config_file).parent_path();
}

}  // namespace

SourceService::SourceService(std::filesystem::path config_file)
    : data_(JsonFileReader().read(config_file)),
      path_resolver_(project_root_of(config_file)) {}

std::vector<Band> SourceService::bands() const {
    std::vector<Band> result;

    for (const auto& band_json : data_.at("bands")) {
        result.push_back(Band{
            band_json.at("id").get<std::string>(),
            band_json.at("name").get<std::string>(),
            band_json.at("logo").get<bool>(),
        });
    }

    return result;
}

std::filesystem::path SourceService::bands_root() const {
    return path_resolver_.resolve(data_.value("path", std::string("bands")));
}

std::vector<Song> SourceService::songs(const Band& band) const {
    std::vector<Song> result;
    const auto bands_root_path = bands_root();

    for (const auto& song_json : data_.at(band.id)) {
        const auto song_id = song_json.at("id").get<std::string>();

        std::vector<Audio> audios;
        for (const auto& audio_json : song_json.at("audios")) {
            const auto file_name = audio_json.at("file").get<std::string>();
            const auto file_path =
                path_resolver_.audio_file_path(bands_root_path, band.id, song_id, file_name);

            audios.push_back(Audio{
                file_name,
                audio_json.at("name").get<std::string>(),
                file_path.string(),
            });
        }

        result.push_back(Song{
            song_id,
            song_json.at("name").get<std::string>(),
            song_json.at("autoforward").get<bool>(),
            std::move(audios),
        });
    }

    return result;
}

void SourceService::validate_audio_files() const {
    std::vector<std::filesystem::path> audio_paths;

    for (const auto& band : bands()) {
        for (const auto& song : songs(band)) {
            for (const auto& audio : song.audios) {
                audio_paths.emplace_back(audio.file);
            }
        }
    }

    validator_.validate(audio_paths);
}

}  // namespace cherry
